### PR TITLE
zstream: add unit tests

### PR DESCRIPTION
**Pull request by GarthSnyder**
This PR adds a unit-testing harness to the `zstream` command along with additional tests for `zstream_queue.`

Tests are run with the (undocumented) `zstream selftest` subcommand:

```
usage: zstream selftest [-l] [-s seed] [-t nthreads] module [test ...]

	-l         list available tests
	-s seed    seed for pseudo-random workloads (for replays)
	-t num     size of the shared worker thread pool

Available modules: queue
```

There are 7 tests within the `queue` group, each of which prepares a workload specification that's then sent to `run_queue_workloads()` for execution. The individual tests exercise a variety of load conditions and edge cases.

I added a test to the `zstream` group to run `zstream selftest queue` and `zstream selftest -t 1 queue`. I see that there's a provision for unit tests in ZTS, but this seemed like a cleaner approach given the dependencies among parts of `zstream`.

The selftest harness includes a watchdog timer set at 120 seconds. If a test times out, the backtrace code in zstream_backtrace.[ch] runs to dump a stack trace of every thread. That backtrace seems to work correctly on all the test VMs.

### Types of Changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [x] Quality assurance (non-breaking change which makes the code more robust against bugs)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [ x All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: cmd/zstream/zstream_queue.c [hunk 1]
```diff
@@ -164,8 +164,11 @@ zstream_queue_set_num_threads(uint_t n)
 {
 	if (pool_initialized) {
 		errx(1, "thread pool size must be set before creating queues");
+	} else if (n == 0) {
+		errx(1, "number of threads must be at least 1");
 	} else if (n < MIN_THREADS) {
-		errx(1, "number of threads must be at least %d", MIN_THREADS);
+		warnx("using only %u threads my limit performance, setting "
+		    "anyway...", n);
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream_queue.c:180]: Fixed

File: cmd/zstream/zstream_selftest.c [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+{
+	(void) sig;
+	const char msg[] = "\nselftest: watchdog timeout in ";
+	if (write(STDERR_FILENO, msg, sizeof (msg) - 1) < 0 ||
+	    write(STDERR_FILENO, current_test, strlen(current_test)) < 0 ||
+	    write(STDERR_FILENO, "\n", 1) < 0) {
+		_exit(1);
+	}
+	_exit(1);
+}
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream_selftest.c]: Fixed.

File: cmd/zstream/zstream_util.c [hunk 3]
```diff
@@ -179,3 +184,29 @@ compress_buffer(uint8_t *inbuff, size_t inbuff_size,
...
+}
+
+void
+pthread_register_self(void)
+{
+	static pthread_once_t init_pthread_tracking = PTHREAD_ONCE_INIT;
+	pthread_once(&init_pthread_tracking, initialize_pthread_key);
+	pthread_setspecific(thread_count_key, (void *)0xdeadbeef);
+	atomic_inc_32(&num_pthreads);
+}
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream_util.c]: Fixed #1 and added a sentinel variable's address instead of 0xdeadbeef. The fix to #2 is "Don't do that."

File: cmd/zstream/zstream_selftest_queue.c [hunk 4]
```diff
@@ -0,0 +1,641 @@
...
+	uint32_t	qi_delay_us;
+	uint64_t	qi_seq;
+	uint64_t	qi_check;
+	size_t		qi_cost;
+	uint32_t	qi_process_count;
+	uint8_t		qi_pattern[];
+} qtest_item_t;
+
+typedef struct {
+	uint32_t	qc_producers;		/* Number of producersl */
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream_selftest_queue.c]: Fixed

File: cmd/zstream/zstream_selftest.h [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+typedef void test_function_f(void);
+
+typedef struct {
+	const char	*tc_name;
+	test_function_f	*tc_func;
+} test_case_t;
+
+/*
+ * Modules with test cases to offer. Each array ends with a NULL sc_name.
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream_selftest.h]: Fixed

File: cmd/zstream/zstream_selftest.c [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+		    sizeof (selftest_seed));
+	(void) printf("Using seed 0x%016jx (replay with -s 0x%jx)\n",
+	    (uintmax_t)selftest_seed, (uintmax_t)selftest_seed);
+
+	if (nthreads > 0)
+		zstream_queue_set_num_threads(nthreads);
+
+	struct sigaction sa = { .sa_handler = watchdog_fire };
+	if (sigaction(SIGALRM, &sa, NULL) != 0)
+		err(1, "failed to install watchdog signal handler");
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream_selftest.c]: Fixed

File: cmd/zstream/zstream_selftest.c [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+		}
+	} else {
+		for (int i = 1; i < argc; i++) {
+			run_case(find_case(module, argv[i]));
+			count++;
+		}
+	}
+	(void) printf("All %d %s selftest%s passed\n", count, module->sm_name,
+	    count == 1 ? "" : "s");
+	return (0);
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream_selftest.c:211]: Fixed

File: cmd/zstream/zstream_util.h [hunk 8]
```diff
@@ -36,6 +36,8 @@ typedef struct {
 	int			cs_level;
 } compression_spec_t;
 
+extern uint32_t num_pthreads;
+
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream_util.h:44]: Fixed

File: cmd/zstream/zstream_selftest_queue.c [hunk 9]
```diff
@@ -0,0 +1,641 @@
...
+	 * the kernel a moment to retire task entries before declaring
+	 * failure.
+	 */
+	for (int i = 0; i < 500; i++) {
+		if (num_pthreads == baseline)
+			return;
+		(void) usleep(2000);
+	}
+	errx(1, "thread pool failed to spin down (%d threads, expected %d)",
+	    num_pthreads, baseline);
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream_selftest_queue.c]: Fixed

File: cmd/zstream/zstream_selftest_queue.c [hunk 10]
```diff
@@ -0,0 +1,641 @@
...
+	selftest_rng_init(&rng, cfg->qc_rng_stream + 1000 + pa->qp_id);
+
+	for (uint64_t seq = 0; seq < cfg->qc_items; seq++) {
+
+		qtest_item_t item_xfer = {
+			.qi_producer = pa->qp_id,
+			.qi_seq = seq,
+			.qi_process_count = 0,
+			.qi_check = item_check_value(pa->qp_id, seq)
+		};
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream_selftest_queue.c:196]: I don't think this is right. The actual standard is paywalled, but the WG14 N1570 draft of 2011/04/12 says: "all subobjects that are not initialized explicitly shall be initialized implicitly the same as
objects that have static storage duration." (section 6.7.9 para 19). Paragraph 10 says, of static storage, "if it [thing being initialized] has arithmetic type, it is initialized to (positive or unsigned) zero."

File: cmd/zstream/zstream_selftest_queue.c [hunk 11]
```diff
@@ -0,0 +1,641 @@
...
+/*
+ * Spin the thread pool up and down repeatedly. After every drain the
+ * process must be back to its baseline thread count. The second phase
+ * races a fresh zstream_queue_create() against the previous queue's
+ * spin-down to exercise the create-vs-spindown locking.
+ */
+static void
+queue_cycles(void)
+{
+	uint32_t baseline = num_pthreads;
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream_selftest_queue.c]: Fixed.

File: cmd/zstream/zstream_selftest.h [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+{
+	return (bound ? selftest_rng_next(rng) % bound : 0);
+}
+
+/*
+ * Number of live threads in this process, or -1 if that can't be
+ * determined on this platform. Used to verify thread pool spin-down.
+ */
+int
+selftest_count_threads(void);
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream_selftest.h]: Woops, fixed.

File: cmd/zstream/zstream_util.c [hunk 13]
```diff
@@ -179,3 +184,32 @@ compress_buffer(uint8_t *inbuff, size_t inbuff_size,
...
+thread_destroyed(void *arg)
+{
+	(void) arg;
+	atomic_dec_32(&num_pthreads);
+}
+
+/*
+ * This key ensures that the thread_destroyed() function is run no matter
+ * how the monitored thread stops.
+ */
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream_util.c]: Fair enough.

File: cmd/zstream/zstream_selftest.c [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+	size_t len = 0;
+	while (current_test[len] != '\0') {
+		len++;
+	}
+	if (write(STDERR_FILENO, msg, sizeof (msg) - 1) < 0 ||
+	    write(STDERR_FILENO, current_test, len) < 0 ||
+	    write(STDERR_FILENO, "\n", 1) < 0) {
+		_exit(1);
+	}
+	_exit(1);
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream_selftest.c]: The test name is set before the test is run, and the watchdog timer is serviced two minutes later, if at all. I don't think there's a legitimate issue here.

**Review by behlendorf — COMMENTED** [review]
Building a `selftest` sub-command in to the utility is a new one for us, but in this instance I think it's pretty reasonable.  This way all of the `zstream` code is self contained.  I suppose we could wrap the selftest functionality with `#ifdef ZFS_DEBUG` guards so it's only enabled in debug builds, but since it's undocumented that seems sufficient.

The CI did seem to trip the watchdog in the last set of runs.  It's not clear to me though if that's a real bug, or just the CI being sluggish.

https://github.com/openzfs/zfs/actions/runs/29962575516/job/89067001181?pr=18829

```
zstream/zstream_selftest_queue_001_pos (54 KiB)
  [2026-07-23T02:18:16.073423] Test: /usr/share/zfs/zfs-tests/tests/functional/zstream/zstream_selftest_queue_001_pos (run as root) [02:01] [FAIL]
  02:16:14.80 ASSERTION: zstream self-tests for zstream_queue all pass
  02:16:14.81 Using seed 0x00415d5997f00adc (replay with -s 0x415d5997f00adc)
  02:16:14.82 Running queue_basic          ... OK
  02:16:14.92 Running queue_edge_cases     ... OK
  02:16:14.93 Running queue_zero_cost      ... OK
  02:16:15.39 Running queue_torture        ... OK
  02:16:15.76 Running queue_multi_producer ... OK
  02:16:15.90 Running queue_multi_queue    ... OK
  02:18:15.91 
  02:18:15.91 selftest: watchdog timeout in queue_cycles
  02:18:15.91 Running queue_cycles         ... ERROR: zstream selftest queue exited 1
```

**Comment by GarthSnyder** [comment]
That's not very informative, is it. Will the VM test harness save core dumps? I can always have the watchdog send a SIGABRT to itself. Meanwhile I'll run overnight with ThreadSanitizer on.

**Comment by behlendorf** [comment]
The CI does collect logs, but I suspect it wouldn't have scooped up the core dump.  You can download what it did gather from the summary page, https://github.com/openzfs/zfs/actions/runs/29962575516?pr=18829 .  Also take a look at `zdb`, `zhack`, and `ztest` they register a signal handler which dumps a backtrace on SIGSEGV and SIGABRT.  You could do the same for `zstream`.

**Comment by behlendorf** [comment]
@GarthSnyder any news on reproducing the selftest failure locally?

**Comment by GarthSnyder** [comment]
I've run some long test sessions, but no luck reproducing anything yet. This would have been my own watchdog timing out after 120 seconds, not the ZTS watchdog, and it happened on a relatively intensive test. But I still suspect an actual issue and I think it's worth adding backtraces. Unfortunately, the existing libspl backtrace code is thread-specific, so I would need to add some infrastructure to signal individual threads in sequnce. Let me switch this to draft status while I work on that.

**Comment by GarthSnyder** [comment]
By progressively reducing the watchdog timer length and pushing updates, I've found that it has to be really short (a few seconds) to have a chance of triggering because of system load or random variation. I think it's more likely than not that the original test failure was the result of a real deadlock. But despite ~10 CI runs and fairly extensive testing on my own systems and VMs, I haven't been able to reproduce any problems.

I have set up all-thread backtraces for the test log in the event of a watchdog timeout, and they seem to be working fine. 

I'd like to prophylactically simplify a couple of things in zstream_queue. Some of the locking complexity comes from the fact that the thread pool terminates threads once there are no active queues. That's not currently serving any useful purpose, and given the one-shot context, I don't think a need for it is likely to show up in the future.

Another potential source of risk is lock-free queue scoring. Well, locks are held, just not the locks on individual queues. Because enqueues are blocked while scoring, there shouldn't be any way for scoring to underestimate demand. But I've also found through profiling that this feature isn't actually increasing performance, so the scoring thread may as well lock the queues. There will still be time skew among queues, but no possibility of misreading any individual queue.

Does this sound reasonable? I think I'm just going to leave this PR as a draft for now and work on getting those simplifications into shape.

File: cmd/zstream/zstream_queue.c [hunk 15]
```diff
@@ -29,7 +29,7 @@
 #include <sys/param.h>
 #include <sys/random.h>
 #include <sys/stdtypes.h>
-#include <unistd.h>
+#include <sys/sysmacros.h>
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream_queue.c:31]: Fixed

**Review by behlendorf — APPROVED** [review]

File: cmd/zstream/zstream_chain.c [hunk 16]
```diff
@@ -265,13 +262,11 @@ zstream_chain_exec(zstream_chain_t chain, chain_attrs_t *attrs)
...
-		    (chain_worker_f *)zstream_chain_worker,
-		    &contexts[i]);
-		VERIFY3S(ret, ==, 0);
-		snprintf(buff, sizeof (buff), "chain-%d", i);
-		pthread_setname_np(worker_threads[i], buff);
+		char name[32];
+		snprintf(name, sizeof (name), "chain-%d", i);
+		worker_threads[i] = safe_create_thread(
+		    (thread_f *)zstream_chain_worker, &contexts[i],
+		    name, B_FALSE);
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream_chain.c]: Fixed

File: cmd/zstream/zstream_util.h [hunk 17]
```diff
@@ -21,8 +21,11 @@
 extern "C" {
 #endif
 
+#include <pthread.h>
+#include <signal.h>
 #include <stddef.h>
 #include <stdlib.h>
+#include <stdint.h>
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream_util.h:29]: Fixed

File: cmd/zstream/zstream.h [hunk 18]
```diff
@@ -17,17 +17,27 @@
...
 #ifdef	__cplusplus
 extern "C" {
 #endif
 
+/*
+ * Signals used by the watchdog timer and zstream_queue. The signal mask
+ * must be set properly before any threads are spawned.
+ */
+#define	WATCHDOG_SIGNAL		SIGALRM
+#define	THREAD_BACKTRACE_SIGNAL	SIGRTMIN
```
[the remark this answers is not in the document]
Comment (GarthSnyder) [reply cmd/zstream/zstream.h:31]: I believe Copilot is mistaken here. The current pthreads implementation does use signals from the RT range, but at the library level, SIGRTMIN hides this from you. It's not actually the first RT signal.

**Comment by GarthSnyder** [comment]
This should be ready to go. There were a couple of failures caused by checking the return value of `pthread_setname_np()`, which seems to have sporadic trouble reading or writing `/proc`. It's still checked and will make multiple attempts, but no error messages now.

**Review by behlendorf — APPROVED** [review]

**Comment by behlendorf** [comment]
Looks good.  Thanks for iterating on this, I'll let you know if the `/proc` crop up in the CI.

**Comment by pcd1193182** [comment]
This PR broke `make checkstyle`; as pointed out by @tsoome on slack, we now see the following error:
```
invalid or missing license boilerplate: cmd/zstream/zstream_backtrace.c
invalid or missing license boilerplate: cmd/zstream/zstream_backtrace.h
invalid or missing license boilerplate: cmd/zstream/zstream_selftest.c
invalid or missing license boilerplate: cmd/zstream/zstream_selftest.h
invalid or missing license boilerplate: cmd/zstream/zstream_selftest_queue.c
invalid or missing license boilerplate: tests/zfs-tests/tests/functional/zstream/zstream_selftest_queue_001_pos.ksh
```

**Comment by GarthSnyder** [comment]
Woops, thanks for the note. Fixed in #18925.
